hybridCarboniteMetadataService/src/main/scala/cromwell/services/metadata/hybridcarbonite/CarbonitedMetadataThawingActor.scala

```diff
@@ -100,34 +110,46 @@ object CarbonitedMetadataThawingActor {
-        processSubworkflowMetadata(intermediate)
-      case other =>
-        throw new RuntimeException(s"Programmer Error: Unexpected BuildWorkflowMetadataJsonAction message of type '${other.getClass.getSimpleName}': $other")
+    def editFor(action: BuildWorkflowMetadataJsonAction, isRootWorkflow: Boolean): ErrorOr[Json] = {
+      val rootOrSubWorkflowJson = if (isRootWorkflow) json.validNel else json.extractSubworkflowMetadata(action.workflowId.toString)
+      action match {
+        case _: GetLogs => rootOrSubWorkflowJson.map(JsonEditor.logs)
+        case _: WorkflowOutputs => rootOrSubWorkflowJson.map(JsonEditor.outputs)
+        case get: GetMetadataAction =>
+          val intermediate = get.key match {
+            case MetadataQuery(_, None, None, None, None, _) => rootOrSubWorkflowJson
+            case MetadataQuery(_, None, Some(key), None, None, _) => rootOrSubWorkflowJson.map(JsonEditor.includeJson(_, NonEmptyList.of(key)))
+            case MetadataQuery(_, None, None, includeKeys, excludeKeys, _) => rootOrSubWorkflowJson.map(JsonEditor.includeExcludeJson(_, includeKeys, excludeKeys))
+            // The following three don't bother with exclusions or inclusions since they are only used internally
+            // and the existing client code is robust to superfluous data.
+            case MetadataQuery(_, Some(_), None, None, None, _) => rootOrSubWorkflowJson
+            case MetadataQuery(_, Some(_), Some(_), None, None, _) => rootOrSubWorkflowJson
+            case MetadataQuery(_, Some(_), None, _, _, _) => rootOrSubWorkflowJson
+            case wrong => throw new RuntimeException(s"Programmer Error: Invalid MetadataQuery: $wrong")
+          }
+          // For carbonited metadata, "expanded" subworkflows translates to not deleting subworkflows out of the root workflow that already
+          // contains them. So `inpJson` for expanded subworkflows and `JsonEditor.replaceSubworkflowMetadataWithId` for unexpanded subworkflows.
+          val processSubworkflowMetadata: ErrorOr[Json] => ErrorOr[Json] = {
+            inpJson => if (get.key.expandSubWorkflows) inpJson else inpJson.flatMap(JsonEditor.replaceSubworkflowMetadataWithId)
+          }
+          processSubworkflowMetadata(intermediate)
+        case other =>
+          throw new RuntimeException(s"Programmer Error: Unexpected BuildWorkflowMetadataJsonAction message of type '${other.getClass.getSimpleName}': $other")
+      }
     }
 
     def updateLabels(labels: Map[WorkflowId, Map[String, String]]): ErrorOr[Json] = JsonEditor.updateLabels(json, labels)
 
+    def extractSubworkflowMetadata(subWorkflowId: String): ErrorOr[Json] = {
+      JsonEditor.extractSubWorkflowsMetadata(subWorkflowId, json).flatMap {
+        case Some(subworkflowMetadata) => subworkflowMetadata.validNel
+        case None => s"Metadata for subworkflow $subWorkflowId was not found".invalidNel
```

Comment:
TOL not sure how difficult it would be to provide context in the error message, but could this explain that the given subworkflowId was unexpectedly not found in the carbonited metadata JSON of its root workflow?

Reply:
I've updated the error message, but I don't see a good/easy way to provide a root workflow id in it.